### PR TITLE
Use JSONRPC for debugger

### PR DESCRIPTION
**Pull request by davidanthoff**
This is ready for review. It requires https://github.com/julia-vscode/JSONRPC.jl/pull/22.

I don't know whether this PR can reasonably be reviewed... Most of the code has been moved to https://github.com/julia-vscode/DebugAdapter.jl, and there is _a lot_ of reorganization mixed with changes in how things are done. Maybe more realistic to just try to use this branch and see whether it works as a review?

Here is a rough outline of what has changed:
- We use JSONRPC for the communication
- Most of the code has been moved to DebugAdapter.jl
- The TypeScript part largely a message forwarding layer now, with some important exceptions around startup and shutdown
- The debug adapter itself now uses a multi task design: there is one task that runs the actual user code, and a second task that handles messages from the extension. The code is a lot simpler this way, and I think this will also make it a lot easier to support debugging of multiple tasks/threading down the road
- The crash reporting is more comprehensive

### CODE REVIEW

**Comment by davidanthoff** [comment]
Actually, I found a number of new bugs, lets wait with reviewing this.

**Comment by davidanthoff** [comment]
Now _really_ ready for review :)

File: src/debugger/juliaDebug.ts [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+	    response.body.supportsEvaluateForHovers = true
+
+	    // make VS Code to show a 'step back' button
+	    response.body.supportsStepBack = false
+
+	    // make VS Code to support data breakpoints
+	    response.body.supportsDataBreakpoints = false
+
+	    // make VS Code to support completion in REPL
+	    response.body.supportsCompletionsRequest = false
```
Comment (pfitzseb) [inline src/debugger/juliaDebug.ts:85]: We can actually have these at some point.

File: src/debugger/juliaDebug.ts [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+	    response.body.supportsCancelRequest = false
+
+	    response.body.supportsTerminateRequest = true
+
+	    // make VS Code send the breakpointLocations request
+	    response.body.supportsBreakpointLocationsRequest = false
+
+	    response.body.supportsConditionalBreakpoints = true
+
+	    response.body.supportsHitConditionalBreakpoints = false
```
Comment (pfitzseb) [inline src/debugger/juliaDebug.ts:98]: Can also have these in the future.

File: src/debugger/juliaDebug.ts [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+	    response.body.supportsLogPoints = false
+
+	    response.body.supportsExceptionInfoRequest = true
+
+	    response.body.supportsRestartFrame = true
+
+	    response.body.supportsSetVariable = true
+
+	    response.body.exceptionBreakpointFilters = [
+	        { filter: 'compilemode', label: 'Enable compile mode (experimental)', default: false },
```
Comment (pfitzseb) [inline src/debugger/juliaDebug.ts:109]:
```suggestion
	        { filter: 'compilemode', label: 'Compiled Mode (experimental)', default: false },
```

File: src/debugger/juliaDebug.ts [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+	        { filter: 'error', label: 'Break any time an uncaught exception is thrown', default: true },
```
Comment (pfitzseb) [inline src/debugger/juliaDebug.ts:110]:
```suggestion
	        { filter: 'error', label: 'Uncaught Exceptions', default: true },
```

File: src/debugger/juliaDebug.ts [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+	        { filter: 'throw', label: 'Break any time a throw is executed', default: false }
```
Comment (pfitzseb) [inline src/debugger/juliaDebug.ts:111]:
```suggestion
	        { filter: 'throw', label: 'All Exceptions', default: false }
```

File: src/debugger/juliaDebug.ts [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+	        }
+	    })
+	    this._debuggeeTerminal.show(false)
+	    const asdf: Array<Disposable> = []
+	    vscode.window.onDidCloseTerminal((terminal) => {
+	        if (terminal === this._debuggeeTerminal) {
+	            this.sendEvent(new TerminatedEvent())
+	            asdf[0].dispose()
+	        }
+	    }, this, asdf)
```
Comment (pfitzseb) [inline src/debugger/juliaDebug.ts:248]:
```suggestion
	    const disposables: Array<Disposable> = []
	    vscode.window.onDidCloseTerminal((terminal) => {
	        if (terminal === this._debuggeeTerminal) {
	            this.sendEvent(new TerminatedEvent())
	            disposables.forEach(d => d.dispose())
	        }
	    }, this, disposables)
```

**Comment by davidanthoff** [comment]
> but that shouldn't hold up merging this.

We first need a review and approval for https://github.com/julia-vscode/JSONRPC.jl/pull/22 :)

**Pull request merged**